### PR TITLE
Fix nullref caused by unknown value to the ignoreCase parameter of Type.GetType

### LINKED ISSUE #1741: Null-reference exception in Mono.Linker.Dataflow.ReflectionMethodBodyScanner.HandleCall
I use the [Z.EntityFramework.Plus.EFCore](https://www.nuget.org/packages/Z.EntityFramework.Plus.EFCore/5.1.11) library and it works great with .NET Core 5, but when I use it in my ready-to-run ARM executable, then the following error is shown:
```
ILLink : error IL1005: Z.Expressions.CodeCompiler.CSharp.ExpressionScope.(String,Type&): Error processing method 'Z.Expressions.CodeCompiler.CSharp.ExpressionScope.(String,Type&)' in assembly 'Z.Expressions.Eval.dll' [/builds/xxx.csproj]
  Mono.Linker.LinkerFatalErrorException: ILLink: error IL1005: Z.Expressions.CodeCompiler.CSharp.ExpressionScope.(String,Type&): Error processing method 'Z.Expressions.CodeCompiler.CSharp.ExpressionScope.(String,Type&)' in assembly 'Z.Expressions.Eval.dll'
   ---> System.NullReferenceException: Object reference not set to an instance of an object.
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.HandleCall(MethodBody callingMethodBody, MethodReference calledMethod, Instruction operation, ValueNodeList methodParams, ValueNode& methodReturnValue)
     at Mono.Linker.Dataflow.MethodBodyScanner.HandleCall(MethodBody callingMethodBody, Instruction operation, Stack`1 currentStack)
     at Mono.Linker.Dataflow.MethodBodyScanner.Scan(MethodBody methodBody)
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.ScanAndProcessReturnValue(MethodBody methodBody)
     at Mono.Linker.Steps.MarkStep.MarkReflectionLikeDependencies(MethodBody body, Boolean requiresReflectionMethodBodyScanner)
     at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger)
```

### PR DESCRIPTION
**Pull request by vitek-karas**
We have special handling for cases where ignoreCase is not constant false (since linker won't be able to correctly mark the required types). This handling was written such that it assumed the value for the parameter was "known" that is we know at least something about it. If it's completely unknown, that's represented as null and the code crashes on it.

Added a test case for this and made the fix.

Fixes #1741

### CODE REVIEW

**Review by tlakollo — APPROVED** [review]
I got stuck thinking how to use the Kind to solve the boolean, looks like checking for parameter boolean works better and is more efficient

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 1]
```diff
@@ -820,8 +820,7 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
 						reflectionContext.AnalyzingPattern ();
 
 						var parameters = calledMethod.Parameters;
-						if ((parameters.Count == 3 && (methodParams[2].Kind == ValueNodeKind.MethodReturn || methodParams[2].Kind == ValueNodeKind.ConstInt || methodParams[2].Kind == ValueNodeKind.LoadField)
-							&& (methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0)) ||
+						if ((parameters.Count == 3 && parameters[2].ParameterType.IsTypeOf ("System", "Boolean") && (methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0)) ||
```
Comment (MichalStrehovsky) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:823]: Nit: check for primitive types can be done more efficiently

```suggestion
						if ((parameters.Count == 3 && parameters[2].ParameterType.MetadataType == MetadataType.Boolean && (methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0)) ||
```

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 2]
```diff
@@ -820,8 +820,7 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
 						reflectionContext.AnalyzingPattern ();
 
 						var parameters = calledMethod.Parameters;
-						if ((parameters.Count == 3 && (methodParams[2].Kind == ValueNodeKind.MethodReturn || methodParams[2].Kind == ValueNodeKind.ConstInt || methodParams[2].Kind == ValueNodeKind.LoadField)
-							&& (methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0)) ||
+						if ((parameters.Count == 3 && parameters[2].ParameterType.MetadataType == MetadataType.Boolean && (methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0)) ||
```
Comment (marek-safar) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:823]: nit: instead of `methodParams[2].AsConstInt () == null || methodParams[2].AsConstInt () != 0` you could add something like `methodParams[2].IsDefault ()`

File: test/Mono.Linker.Tests.Cases/Reflection/TypeUsedViaReflection.cs [hunk 3]
```diff
@@ -359,5 +361,23 @@ static Type GetTypeFromAssembly (Assembly assembly, string name, bool caseSensit
 		{
 			return assembly == null ? Type.GetType (name, caseSensitive) : assembly.GetType (name, caseSensitive);
 		}
+
+		[Kept]
+		[ExpectedWarning ("IL2096", "'System.Type.GetType(String,Boolean,Boolean)'")]
+		static void TestUnkownIgnoreCase3Params (int num)
```
Comment (marek-safar) [inline test/Mono.Linker.Tests.Cases/Reflection/TypeUsedViaReflection.cs:367]: Do we have coverage for similar cases for other intrinsics?

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs (hunk 2 above)
Comment (vitek-karas) [reply src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:823]: While technically true I'm not sure it's worth it. The case here is specific - the parameter is a boolean and we're looking for `true` or `unknown`. Neither is "default" semantically.

Comment (vitek-karas) [reply src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:823]: Actually it can be simplified as simple `!=0` - or the boolean equivalent `!=false`.

File: test/Mono.Linker.Tests.Cases/Reflection/TypeUsedViaReflection.cs (hunk 3 above)
Comment (vitek-karas) [reply test/Mono.Linker.Tests.Cases/Reflection/TypeUsedViaReflection.cs:367]: I checked all other cases where we consume the "method parameter" value in the intrinsic handling and none has this problem. We probably don't have coverage though. There are lot of places where we consume binding flags which could all be "null" - and the code has handling for it.

I'm adding a test for the null case for binding flags in one place - but all the other places use the same helper to handle this, so in terms of code coverage it's good. Writing one for each case would be lot of new test cases.

**Review by marek-safar — APPROVED** [review]

**Comment by marek-safar** [comment]
We should consider backporting the fix to 5.0

**Comment by vitek-karas** [comment]
I ported this to 5.0 as https://github.com/mono/linker/pull/1759
@marek-safar will you take it to triage?

**Comment by marek-safar** [comment]
> will you take it to triage?

Yep
